test/e2e/apimachinery/flowcontrol.go

```diff
@@ -0,0 +1,137 @@
+			context.TODO(),
+			&flowcontrolv1alpha1.PriorityLevelConfiguration{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: testingPriorityLevelName,
+				},
+				Spec: flowcontrolv1alpha1.PriorityLevelConfigurationSpec{
+					Type: flowcontrolv1alpha1.PriorityLevelEnablementLimited,
+					Limited: &flowcontrolv1alpha1.LimitedPriorityLevelConfiguration{
+						AssuredConcurrencyShares: 1, // will have at minimum 1 concurrency share
+						LimitResponse: flowcontrolv1alpha1.LimitResponse{
+							Type: flowcontrolv1alpha1.LimitResponseTypeReject,
+						},
+					},
+				},
+			},
+			metav1.CreateOptions{})
+		framework.ExpectNoError(err)
+
+		defer func() {
+			// clean-ups
+			err := f.ClientSet.FlowcontrolV1alpha1().PriorityLevelConfigurations().Delete(context.TODO(), testingPriorityLevelName, metav1.DeleteOptions{})
+			framework.ExpectNoError(err)
+			err = f.ClientSet.FlowcontrolV1alpha1().FlowSchemas().Delete(context.TODO(), testingFlowSchemaName, metav1.DeleteOptions{})
+			framework.ExpectNoError(err)
+		}()
+
+		ginkgo.By("creating a testing flowschema")
+		createdFlowSchema, err := f.ClientSet.FlowcontrolV1alpha1().FlowSchemas().Create(
+			context.TODO(),
+			&flowcontrolv1alpha1.FlowSchema{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: testingFlowSchemaName,
+				},
+				Spec: flowcontrolv1alpha1.FlowSchemaSpec{
+					MatchingPrecedence: 1000, // a rather higher precedence to ensure it make effect
+					PriorityLevelConfiguration: flowcontrolv1alpha1.PriorityLevelConfigurationReference{
+						Name: testingPriorityLevelName,
+					},
+					DistinguisherMethod: &flowcontrolv1alpha1.FlowDistinguisherMethod{
+						Type: flowcontrolv1alpha1.FlowDistinguisherMethodByUserType,
+					},
+					Rules: []flowcontrolv1alpha1.PolicyRulesWithSubjects{
+						{
+							Subjects: []flowcontrolv1alpha1.Subject{
+								{
+									Kind: flowcontrolv1alpha1.SubjectKindUser,
+									User: &flowcontrolv1alpha1.UserSubject{
+										Name: matchingUsername,
+									},
+								},
+							},
+							NonResourceRules: []flowcontrolv1alpha1.NonResourcePolicyRule{
+								{
+									Verbs:           []string{flowcontrolv1alpha1.VerbAll},
+									NonResourceURLs: []string{flowcontrolv1alpha1.NonResourceAll},
+								},
+							},
+						},
+					},
+				},
+			},
+			metav1.CreateOptions{})
+		framework.ExpectNoError(err)
+
+		ginkgo.By("response headers should contain flow-schema/priority-level uid")
+
+		if !testResponseHeaderMatches(f, matchingUsername, string(createdPriorityLevel.UID), string(createdFlowSchema.UID)) {
+			framework.Failf("matching user doesnt received UID for the testing priority-level and flow-schema")
+		}
+		if testResponseHeaderMatches(f, nonMatchingUsername, string(createdPriorityLevel.UID), string(createdPriorityLevel.UID)) {
+			framework.Failf("non-matching user unexpectedly received UID for the testing priority-level and flow-schema")
+		}
```

Comment:
Nit: if we decompose the `testResponseHeaderMatches` function, we can narrow it down to which header does not match.

```
{
        ginkgo.By("invoke API impersonating the matching user")
 
        response, err := invokeAPI(f, matchingUsername)
	if  err != nil {
		framework.Failf("Failed to invoke API with impersonaion")
	}

       ginkgo.By("response headers should contain flow-schema/priority-level uid for matching user")
       framework.ExpectEqual(fsUID, getMatchedFlowSchemaUID(response), "FlowSchema UID does not match")
       framework.ExpectEqual(plUID, getMatchedPriorityLevelConfigurationUID(response), "PriorityLevelConfiguration  UID does not match")       

      ginkgo.By("invoke API impersonating the non-matching user")
      //  non matching follows here.
}

func invokeAPI(f *framework.Framework, impersonatingUser string) (http.Response, error) {
	config := rest.CopyConfig(f.ClientConfig())
	config.Impersonate.UserName = impersonatingUser
	roundTripper, err := rest.TransportFor(config)
	framework.ExpectNoError(err)

	req, err := http.NewRequest(http.MethodGet, f.ClientSet.CoreV1().RESTClient().Get().AbsPath("version").URL().String(), nil)	
        framework.ExpectNoError(err)

	response, err := roundTripper.RoundTrip(req)
	framework.ExpectNoError(err)

        return response, nil
}

func getMatchedFlowSchemaUID(response *http.Response) string {
        return response.Header.Get(flowcontrolv1alpha1.ResponseHeaderMatchedFlowSchemaUID)
}

func getMatchedPriorityLevelConfigurationUID(response *http.Response) string {
        return response.Header.Get(flowcontrolv1alpha1.ResponseHeaderMatchedPriorityLevelConfigurationUID)
}